[Hyun] step-5 구현 : 로그인 체크 인터셉터, 게시글 작성, 수정, 삭제 권한 체크 구현

ghkdgus29 opened the pull request:
세션에 로그인한 유저의 PK( int )값만 저장하도록 하였기 때문에 게시글에 로그인한 유저의 userId( String ) 을 출력하기 위해서 리포지토리 단에서 join 을 하였습니다.
이때 join 을 사용하여 Article 도메인 내의 writer 필드에 userId 를 입력하게 됩니다.

제가 구현한 것처럼 join을 사용하는 것이 나은 방법일지, 세션에 User객체를 넣고 Article 도메인 생성시에 필요한 user의 pk와 userId를 넣는것이 나을지 고민하였습니다. 
성능상은 DB에 부하가 없는 후자가 나을 수 있겠으나, 이런 방식으로 구현하면 Article 컨트롤러가 User 도메인에 의존하게 되는데, 이보다는 int값인 userId를 세션에서 꺼내 사용하는게 낫지 않을까 고민했습니다.

필연적으로 많은 폼을 건드리게 되었는데 헷갈리고 잔실수가 많이 나오는 과정이었습니다.
나중에 협업 프로젝트를 수행하면 이러한 폼 수정에서 해방될 거 같아 벌써 기대됩니다.

ghkdgus29 committed:
feat: 로그인시 발생하는 예외를 처리하는 공통 관심사를 분리

@ExceptionHandler 를 사용하여 로그인시 발생하는 IllegalArgumentException 을 잡아 에러페이지를 출력하도록 수정

ghkdgus29 committed:
feat: 로그인 여부를 체크하는 로그인 인터셉터 추가

로그인하지 않은 사용자는 글작성 및 글 세부내용을 볼 수 없다
해당 동작을 수행하려고 하면 로그인 페이지로 리디렉션 처리한다.

ghkdgus29 committed:
feat: article 테이블에 writer 속성을 제거하고 user의 pk값을 fk 로 갖도록 수정

이에 따라 컨틀롤러, 테이블, 테스트, 폼을 모두 수정하였다.
컨트롤러의 경우 질문작성폼에서 넘어오는 정보가 2개밖에 안되기 때문에 @ModelAttribute를 제거하고 @RequestParam 으로 직접 받아오도록 변경하였다.
세션에 존재하는 회원의 id 값을 Article 객체를 생성할 때 넣어준다.

ghkdgus29 committed:
feat: Article 객체 내에 글쓴 회원이름을 나타내는 writer 필드를 추가

writer 필드는 Article을 조회해올 때, userId를 사용하여 테이블 조인하는 시점에 저장된다.

ghkdgus29 committed:
feat: Article 저장소에서 id 를 통해 삭제하는 기능 구현

ghkdgus29 committed:
feat: 로그인 체크 인터셉터에 회원 상세정보 URL 도 추가

ghkdgus29 committed:
feat: 컨트롤러 전역에서 발생하는 IllegalArgumentException 예외를 잡아 에러페이지를 출력하도록 구조 변경

ghkdgus29 committed:
feat: 게시글 삭제 기능 구현

만약 로그인한 회원과 게시글 작성자가 다르면 삭제할 수 없다

ghkdgus29 committed:
feat: Article 저장소에 update 기능 구현

게시글 제목과 내용을 수정 가능

ghkdgus29 committed:
feat: 게시글 수정 기능 구현

updateForm 에서는 게시글 수정이므로 PUT 요청을 하도록 변경
로그인 유저와 게시글 작성자가 다른 경우 수정 불가

honux77 commented on src/main/java/kr/codesqaud/cafe/config/WebConfig.java:
```diff
@@ -0,0 +1,18 @@
+    @Override
+    public void addInterceptors(InterceptorRegistry registry) {
+        registry.addInterceptor(new LoginInterceptor())
+                .order(1)
```
order 에 1이 아니라 상수를 쓸 수 있지 않나요? ORDER.XXXX

honux77 replied:
안 쓰고 할 수도 있습니다만 인터셉터로 로그인을 구현하는 시도를 한 건 좋은 것 같아요.

honux77 commented on src/main/java/kr/codesqaud/cafe/repository/H2DBArticleRepository.java:
```diff
@@ -37,7 +38,7 @@ public void save(Article article) {
     @Override
     public Article findById(int id) {
-        String sql = "select id, writer, title, contents, createDate from article where id=:id";
+        String sql = "select a.*, u.user_id as writer from article a join users u on a.user_id=u.id where a.id=:id";
```
여기에는 사실 join이 필요하지 않겠군요. 그리고 select 쿼리에 *는 사용하지 않는 것이 좋습니다.

honux77 commented on src/main/java/kr/codesqaud/cafe/controller/ArticleController.java:
```diff
@@ -43,17 +42,55 @@ public String showQuestionForm() {
+        Article article = articleRepository.findById(articleIndex);
+
+        if (!loginUser.getUserId().equals(article.getWriter())) {
+            throw new IllegalArgumentException("[ERROR] 자신이 작성하지 않은 게시물은 " + action + "할 수 없습니다.");
```
ㅎㅎ 참심한데 스타일이 뭔가 좀 아쉽네요. 더 좋은 방법은 없을까요?

ghkdgus29 replied:
스타일이라 함은 메서드 시그니처의 가독성이 아쉽다는 말씀이실까요?

honux77 replied:
`throw new IllegalArgumentException("[ERROR] 자신이 작성하지 않은 게시물은 " + action + "할 수 없습니다.");` 이 부분 이야기였어요.

honux77 reviewed (approved):
DB의 Join은 엄청 비싼 연산입니다. 세션에 User를 들고 있는 쪽이 더 낫습니다.
그리고 삭제에는 보통은 soft delete 라는 걸 사용합니다. 지금도 상관은 없지만 관련내용 학습해도 좋을 것 같아요.

👍

honux77 commented on src/main/java/kr/codesqaud/cafe/controller/ArticleController.java:
```diff
@@ -43,17 +42,55 @@ public String showQuestionForm() {
+        return "redirect:/articles/{index}";
+    }
+
+    private void checkUserEqualsWriter(int articleIndex, HttpSession session, String action) {
```
article 테이블에 writer 가 있는 건 좋은 선택이 아닙니다.
외래키로 user 테이블의 pk 만 가지고 있는게 바람직합니다.

honux77 replied:
그리고 테이블 스키마가 없어서 제가 확인하기가 좀 어렵네요. 소스 어딘가에 테이블 작성 SQL도 있으면 좋을 것 같습니다.

ghkdgus29 replied:
테스트 내의 resource 디렉토리 최상단에 위치하는 test-schema.sql 이 실제 DB의 테이블과 동일합니다!!

```sql
create table users
(
    id       int         primary key auto_increment ,
    user_id  varchar(30) not null unique,
    password varchar(30) not null,
    name     varchar(30) not null,
    email    varchar(30) not null
);

create table article
(
    id         int         primary key auto_increment,
    title      varchar(30) not null,
    contents   varchar(50) not null,
    createDate timestamp,
    user_id    int,
    foreign key (user_id) references users (id)
)

```

ghkdgus29 replied:
article 테이블이 외래키로 갖고 있는것은 users 의 id(int)  값입니다. 
만약 article 객체가 writer를 갖고 있지 않게 하려면 모든 게시글을 불러오는 경우에는 join 문을 피하기 어려울것 같은데 이 경우엔 join 을 사용하는 것이 괜찮은 방법일까요?

ghkdgus29 replied:
article 테이블은 writer를 갖고 있지 않고 객체만 갖고 있습니다.

이 부분은 혼동을 줄 수 있으니 수정하는 것이 바람직할 것 같습니다.

honux77 reviewed (changes requested):
Article 객체와 테이블에서 writer를 삭제하고 외래키를 지정해서 수정해 주세요.

ghkdgus29 committed:
fix: 3/23 호눅스 피드백 반영

1. 인터셉터의 order 에서 매직넘버대신 상수 사용하기
-> Ordered.HIGHEST_PRECEDENCE 상수를 사용

2. 게시글 단건 조회에서는 join 사용하지 않기 + select 절 * 사용하지 않기
-> 단건 조회시엔 이중쿼리를 사용하여 join의 사용을 피함

3. Article 객체와 테이블에서 writer 제거
-> 해당 속성들을 제거하고 ArticleWithWriter DTO를 만들어 writer를 표현하도록 변경

ghkdgus29 committed:
feat: 캐시무효화 인터셉터 추가

게시글 삭제후 뒤로가기를 누르면 삭제된 게시글이 그대로 보이는 등의 오류가 있어 이를 막기 위해 캐시무효화 인터셉터를 추가
